### PR TITLE
PYD-92: Add custom type docs for Annotated and TypeAliasType

### PR DESCRIPTION
**Pull request by adriangb**
Selected Reviewer: @Kludex

### CODE REVIEW

**Comment by adriangb** [comment]
please review

File: docs/usage/types/custom.md [hunk 1]
```diff
@@ -2,6 +2,283 @@
...
+
+try:
+    ta.validate_python({'x': object()})
+except ValidationError as exc:
+    print(exc)
+    """
+    1 validation error for function-wrap[json_custom_error_validator()]
+      Input is not valid json [type=invalid_json, input_value={'x': <object object at 0x0123456789ab>}, input_type=dict]
+    """
+```
```
Comment (adriangb) [inline docs/usage/types/custom.md:280]: @JelleZijlstra you might find this section interesting, also thank you for your work on this!

File: docs/usage/types/custom.md [hunk 2]
```diff
@@ -2,6 +2,317 @@
 
 You can also define your own custom data types. There are several ways to achieve it.
 
+### Composing types via `Annotated`
+
+[PEP 593] introduced `Annotated` as a way to attach runtime metadata to types without changing how type checkers interpret them.
+Pydantic takes advantage of this to allow you to create types that are identical to the original type as far as type checkers are concerned but add validation, serialize differently, etc.
```
Comment (tpdorsey) [inline docs/usage/types/custom.md:8]:
```suggestion
Pydantic takes advantage of this to allow you to create types that are identical to the original type as far as type checkers are concerned, but add validation, serialize differently, etc.
```

File: docs/usage/types/custom.md [hunk 3]
```diff
@@ -2,6 +2,317 @@
...
+    ta.validate_python(-1)
+except ValidationError as exc:
+    print(exc)
+    """
+    1 validation error for constrained-int
+      Input should be greater than 0 [type=greater_than, input_value=-1, input_type=int]
+    """
+```
+
+Note that you can also use constraints from [`annotated-types`] to make this Pydantic agnostic:
```
Comment (tpdorsey) [inline docs/usage/types/custom.md:34]:
```suggestion
Note that you can also use constraints from [`annotated-types`] to make this Pydantic-agnostic:
```

File: docs/usage/types/custom.md [hunk 4]
```diff
@@ -2,6 +2,317 @@
...
+    print(exc)
+    """
+    1 validation error for constrained-int
+      Input should be greater than 0 [type=greater_than, input_value=-1, input_type=int]
+    """
+```
+
+#### Adding validation and serialization
+
+You can add or override validation, serialization and json schemas to an arbitrary type using the markers that Pydantic exports:
```
Comment (tpdorsey) [inline docs/usage/types/custom.md:61]:
```suggestion
You can add or override validation, serialization, and json schemas to an arbitrary type using the markers that Pydantic exports:
```

File: docs/usage/types/custom.md [hunk 5]
```diff
@@ -2,6 +2,317 @@
...
+    0
+      Input should be greater than 0 [type=greater_than, input_value=-1, input_type=int]
+    """
+```
+
+### Named type aliases
+
+The above examples make use of implicit type aliases.
+This means that they will not be able to have a `title` in JSON schemas and their schema will be copied between fields.
+You can use [PEP 695]'s `TypeAliasType` via it's [typing-extensions] backport to make named aliases, allowing you to define a new type without creating subclasses.
```
Comment (tpdorsey) [inline docs/usage/types/custom.md:154]:
```suggestion
You can use [PEP 695]'s `TypeAliasType` via its [typing-extensions] backport to make named aliases, allowing you to define a new type without creating subclasses.
```

**Review by tpdorsey — APPROVED** [review]
Minor suggested edits. Looks good.

**Pull request merged**